src/metabase/sync/util.clj

```diff
@@ -0,0 +1,231 @@
+(ns metabase.sync.util
+  "Utility functions and macros to abstract away some common patterns and operations across the sync processes, such as logging start/end messages."
+  (:require [clojure.math.numeric-tower :as math]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [metabase
+             [driver :as driver]
+             [events :as events]
+             [util :as u]]
+            [metabase.models.table :refer [Table]]
+            [metabase.query-processor.interface :as qpi]
+            [metabase.sync.interface :as i]
+            [toucan.db :as db]))
+
+;;; +------------------------------------------------------------------------------------------------------------------------+
+;;; |                                              SYNC OPERATION "MIDDLEWARE"                                               |
+;;; +------------------------------------------------------------------------------------------------------------------------+
+
+;; When using the `sync-operation` macro below the BODY of the macro will be executed in the context of several different
+;; functions below that do things like prevent duplicate operations from being ran simultaneously and taking care of
+;; things like event publishing, error handling, and logging.
+;;
+;; These basically operate in a middleware pattern, where the various different steps take a function, and return a new function
+;; that will execute the original in whatever context or with whatever side effects appropriate for that step.
+
+
+;; This looks something like {:sync #{1 2}, :cache #{2 3}} when populated.
+;; Key is a type of sync operation, e.g. `:sync` or `:cache`; vals are sets of DB IDs undergoing that operation.
+(defonce ^:private operation->db-ids (atom {}))
```

Comment:
I would prefer something like this be done via the DB. This is digging us deeper in the hole of "things that prevent us from being multi-server"